tests/subsys/bootloader/bl_validation_neg/src/main.c

```diff
@@ -0,0 +1,147 @@
+/*
+ * Copyright (c) 2018 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
+#include <ztest.h>
+// #include <bl_validation.h>
+#include <fw_info.h>
+#include <pm_config.h>
+// #include <sys/util.h>
+#include <nrfx_nvmc.h>
+#include <linker/linker-defs.h>
+// #include <devicetree.h>
+#include <power/reboot.h>
+#include <bl_storage.h>
+#include <fw_info.h>
+
+
+
+/* These symbols are defined in linker scripts. */
+// extern const u32_t _flash_used[];
+// extern const struct fw_info _firmware_info_start[];
+extern const u32_t _ext_apis_size[];
+extern const u32_t _ext_apis_req_size[];
+// extern const u32_t _fw_info_images_start[];
+// extern const u32_t _fw_info_images_size[];
+// extern const u32_t _fw_info_size[];
+
+extern const struct fw_info m_firmware_info;
+
+u32_t val_info_buf[0x1000];
+
+void test_validation_neg1(void)
+{
+	u32_t copy_len = (u32_t)_flash_used;
+
+	/* Round up to at least the next SPU region. */
+	u32_t new_addr = ROUND_UP(PM_ADDRESS + (PM_SIZE / 2), 0x8000);
+
+	const struct fw_info s1_info = {
+		.magic = {FIRMWARE_INFO_MAGIC},
+		.total_size = PM_S0_ADDRESS - PM_S1_ADDRESS,
+		.size = ((u32_t)_flash_used),
+		.version = CONFIG_FW_INFO_FIRMWARE_VERSION + 1,
+		.address = new_addr,
+		.boot_address = new_addr,
+		.valid = CONFIG_FW_INFO_VALID_VAL,
+		.reserved = {0, 0, 0, 0},
+		.ext_api_num = 0,
+		.ext_api_request_num = 0,
+	};
+
+	const struct fw_info *s1_info_copied = fw_info_find(PM_S1_ADDRESS);
+
+	if (s1_info_copied) {
+		zassert_not_equal(CONFIG_FW_INFO_VALID_VAL,
+			s1_info_copied->valid, "Failed to invalidate S1.\r\n");
+		zassert_equal((u32_t)s1_info_copied, PM_S1_ADDRESS,
+			"S1 info found at wrong address.\r\n");
+		u32_t ret = nrfx_nvmc_page_erase(PM_S1_ADDRESS);
+
+		zassert_equal(NRFX_SUCCESS, ret, "Erase failed.\r\n");
+	} else {
+		for (u32_t erase_addr = new_addr;
+			erase_addr < (new_addr + copy_len);
+			erase_addr += DT_PROP(DT_CHOSEN(zephyr_flash),
+							erase_block_size)) {
+			u32_t ret = nrfx_nvmc_page_erase(new_addr);
+
+			zassert_equal(NRFX_SUCCESS, ret, "Erase failed.\r\n");
+		}
+		nrfx_nvmc_words_write(new_addr, (const u32_t *)PM_ADDRESS,
```

Comment:
Some comments on what is going on would be appreciated.

Reply:
Done